[c10d] sequence number in process group

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55718 [c10d] Increment sequence numbers on collectives.
* **#55319 [c10d] sequence number in process group**

Adds a sequence number class as well as integration with ProcessGroup (nccl and gloo) as part of better debugability.

The main use case is that each ProcessGroup instantiated will have a sequence number initially set by rank 0, and broadcasted to all others. We will increment the number on each collective, thus allowing us to match the numbers appropriately when checking for desynchronization.

This PR just adds the bare-bones integration and verifies sequence numbers are set appropriately at the beginning. The next diff in this stack will increment the sequence number appropriately on collectives.

Note that we initialize sequence numbers in python in `new_group` and `init_process_group` APIs for NCCL and Gloo backends only. The reason we do this instead of the constructor is because the pg constructor in C++ is not a synchronization point, and can result in timeout issues.

Differential Revision: [D27562769](https://our.internmc.facebook.com/intern/diff/D27562769/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27562769/)!

rohan-varma committed:
[c10d] sequence number in process group

Adds a sequence number class as well as integration with ProcessGroup (nccl and gloo) as part of better debugability.

The main use case is that each ProcessGroup instantiated will have a sequence number initially set by rank 0, and broadcasted to all others. We will increment the number on each collective, thus allowing us to match the numbers appropriately when checking for desynchronization.

This PR just adds the bare-bones integration and verifies sequence numbers are set appropriately at the beginning.

Differential Revision: [D27562769](https://our.internmc.facebook.com/intern/diff/D27562769/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27562769/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] sequence number in process group"


Adds a sequence number class as well as integration with ProcessGroup (nccl and gloo) as part of better debugability.

The main use case is that each ProcessGroup instantiated will have a sequence number initially set by rank 0, and broadcasted to all others. We will increment the number on each collective, thus allowing us to match the numbers appropriately when checking for desynchronization.

This PR just adds the bare-bones integration and verifies sequence numbers are set appropriately at the beginning. The next diff in this stack will increment the sequence number appropriately on collectives.

Differential Revision: [D27562769](https://our.internmc.facebook.com/intern/diff/D27562769/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27562769/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] sequence number in process group"


Adds a sequence number class as well as integration with ProcessGroup (nccl and gloo) as part of better debugability.

The main use case is that each ProcessGroup instantiated will have a sequence number initially set by rank 0, and broadcasted to all others. We will increment the number on each collective, thus allowing us to match the numbers appropriately when checking for desynchronization.

This PR just adds the bare-bones integration and verifies sequence numbers are set appropriately at the beginning. The next diff in this stack will increment the sequence number appropriately on collectives.

Differential Revision: [D27562769](https://our.internmc.facebook.com/intern/diff/D27562769/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27562769/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] sequence number in process group"


Adds a sequence number class as well as integration with ProcessGroup (nccl and gloo) as part of better debugability.

The main use case is that each ProcessGroup instantiated will have a sequence number initially set by rank 0, and broadcasted to all others. We will increment the number on each collective, thus allowing us to match the numbers appropriately when checking for desynchronization.

This PR just adds the bare-bones integration and verifies sequence numbers are set appropriately at the beginning. The next diff in this stack will increment the sequence number appropriately on collectives.

Differential Revision: [D27562769](https://our.internmc.facebook.com/intern/diff/D27562769/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27562769/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] sequence number in process group"


Adds a sequence number class as well as integration with ProcessGroup (nccl and gloo) as part of better debugability.

The main use case is that each ProcessGroup instantiated will have a sequence number initially set by rank 0, and broadcasted to all others. We will increment the number on each collective, thus allowing us to match the numbers appropriately when checking for desynchronization.

This PR just adds the bare-bones integration and verifies sequence numbers are set appropriately at the beginning. The next diff in this stack will increment the sequence number appropriately on collectives.

Differential Revision: [D27562769](https://our.internmc.facebook.com/intern/diff/D27562769/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27562769/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] sequence number in process group"


Adds a sequence number class as well as integration with ProcessGroup (nccl and gloo) as part of better debugability.

The main use case is that each ProcessGroup instantiated will have a sequence number initially set by rank 0, and broadcasted to all others. We will increment the number on each collective, thus allowing us to match the numbers appropriately when checking for desynchronization.

This PR just adds the bare-bones integration and verifies sequence numbers are set appropriately at the beginning. The next diff in this stack will increment the sequence number appropriately on collectives.

Differential Revision: [D27562769](https://our.internmc.facebook.com/intern/diff/D27562769/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27562769/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] sequence number in process group"


Adds a sequence number class as well as integration with ProcessGroup (nccl and gloo) as part of better debugability.

The main use case is that each ProcessGroup instantiated will have a sequence number initially set by rank 0, and broadcasted to all others. We will increment the number on each collective, thus allowing us to match the numbers appropriately when checking for desynchronization.

This PR just adds the bare-bones integration and verifies sequence numbers are set appropriately at the beginning. The next diff in this stack will increment the sequence number appropriately on collectives.

Differential Revision: [D27562769](https://our.internmc.facebook.com/intern/diff/D27562769/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27562769/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] sequence number in process group"


Adds a sequence number class as well as integration with ProcessGroup (nccl and gloo) as part of better debugability.

The main use case is that each ProcessGroup instantiated will have a sequence number initially set by rank 0, and broadcasted to all others. We will increment the number on each collective, thus allowing us to match the numbers appropriately when checking for desynchronization.

This PR just adds the bare-bones integration and verifies sequence numbers are set appropriately at the beginning. The next diff in this stack will increment the sequence number appropriately on collectives.

Differential Revision: [D27562769](https://our.internmc.facebook.com/intern/diff/D27562769/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27562769/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] sequence number in process group"


Adds a sequence number class as well as integration with ProcessGroup (nccl and gloo) as part of better debugability.

The main use case is that each ProcessGroup instantiated will have a sequence number initially set by rank 0, and broadcasted to all others. We will increment the number on each collective, thus allowing us to match the numbers appropriately when checking for desynchronization.

This PR just adds the bare-bones integration and verifies sequence numbers are set appropriately at the beginning. The next diff in this stack will increment the sequence number appropriately on collectives.

Differential Revision: [D27562769](https://our.internmc.facebook.com/intern/diff/D27562769/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27562769/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] sequence number in process group"


Adds a sequence number class as well as integration with ProcessGroup (nccl and gloo) as part of better debugability.

The main use case is that each ProcessGroup instantiated will have a sequence number initially set by rank 0, and broadcasted to all others. We will increment the number on each collective, thus allowing us to match the numbers appropriately when checking for desynchronization.

This PR just adds the bare-bones integration and verifies sequence numbers are set appropriately at the beginning. The next diff in this stack will increment the sequence number appropriately on collectives.

Note that we initialize sequence numbers in python in `new_group` and `init_process_group` APIs for NCCL and Gloo backends only. The reason we do this instead of the constructor is because the pg constructor in C++ is not a synchronization point, and can result in timeout issues.

Differential Revision: [D27562769](https://our.internmc.facebook.com/intern/diff/D27562769/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27562769/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] sequence number in process group"


Adds a sequence number class as well as integration with ProcessGroup (nccl and gloo) as part of better debugability.

The main use case is that each ProcessGroup instantiated will have a sequence number initially set by rank 0, and broadcasted to all others. We will increment the number on each collective, thus allowing us to match the numbers appropriately when checking for desynchronization.

This PR just adds the bare-bones integration and verifies sequence numbers are set appropriately at the beginning. The next diff in this stack will increment the sequence number appropriately on collectives.

Note that we initialize sequence numbers in python in `new_group` and `init_process_group` APIs for NCCL and Gloo backends only. The reason we do this instead of the constructor is because the pg constructor in C++ is not a synchronization point, and can result in timeout issues.

Differential Revision: [D27562769](https://our.internmc.facebook.com/intern/diff/D27562769/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27562769/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] sequence number in process group"


Adds a sequence number class as well as integration with ProcessGroup (nccl and gloo) as part of better debugability.

The main use case is that each ProcessGroup instantiated will have a sequence number initially set by rank 0, and broadcasted to all others. We will increment the number on each collective, thus allowing us to match the numbers appropriately when checking for desynchronization.

This PR just adds the bare-bones integration and verifies sequence numbers are set appropriately at the beginning. The next diff in this stack will increment the sequence number appropriately on collectives.

Note that we initialize sequence numbers in python in `new_group` and `init_process_group` APIs for NCCL and Gloo backends only. The reason we do this instead of the constructor is because the pg constructor in C++ is not a synchronization point, and can result in timeout issues.

Differential Revision: [D27562769](https://our.internmc.facebook.com/intern/diff/D27562769/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27562769/)!

[ghstack-poisoned]

SciPioneer commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -2849,4 +2849,29 @@ void ProcessGroupGloo::monitoredBarrier(
+void ProcessGroupGloo::setSequenceNumberForGroup() {
+  if (rank_ == 0) {
+    // Create and broadcast sequence number
+    auto seq = 1 + rand();
```
Add a seed here to make the sequence number deterministic?

rohan-varma replied:
Do we need determinism here? As I understand it, doesn't really matter what the # is, but we just need to increment it and keep it synchronized across processes. And for tests we just validate this behavior andn not the actual number?

SciPioneer replied:
Can you explain why this initial seq number cannot be simply 0?

rohan-varma replied:
Starting it at `0` might be tricky because there could be multiple process groups active within the application (maybe 1 NCCL and 1 Gloo pg). For keeping track of desynchronizations, this would still actually be fine. Although, for debug logs I'd prefer to start them off at different values so it's easy to keep track of which sequence of numbers belongs to which PG. What do you think?

SciPioneer replied:
I think a random value in your implementation is totally fine.

Sometimes I just feel that it will be nice to know at which round a desync occurs. This can give the user a feeling how long does it take to hit the desync, e.g., at the very beginning? or took quite a while? or did different desync failures occur at around the same time?

rohan-varma replied:
This is a good point, I think with the logging we will add with ProcessGroupWrapper, we will report:
1. sequence number
2. name and args of collective call

which hopefully helps make it a bit more clear when the desync happens. In addition, by logging the sequence no., we will be able to have a good idea as to how many collectives occured before the desync.

jiayisuse replied:
> Starting it at `0` might be tricky because there could be multiple process groups active within the application (maybe 1 NCCL and 1 Gloo pg). For keeping track of desynchronizations, this would still actually be fine. Although, for debug logs I'd prefer to start them off at different values so it's easy to keep track of which sequence of numbers belongs to which PG. What do you think?
> 
> > how can we know at which round a desync occurs?
> 
> My plan is to increment the counter on each collective. Since to stay in sync every process should call the collectives the same no. of times in the same order, we will just communicate the sequence number (in detail debug mode) via the store and verify they are consistent across processes. This will help catch issues where a process "skips" a collective.

* NCCL kernel is launched asynchronously. CPU counter won't work to detect desync in most cases.
* Even to start from random seq, there is still the chance to conflict between different pgs. We can assign pg id to include the id in logs. By that mean we can just start form 0 and have good idea of how many collectives are called or finished.

SciPioneer commented on torch/lib/c10d/sequence_num.cpp:
```diff
@@ -0,0 +1,65 @@
+namespace c10d {
+SequenceNum::SequenceNum() : num_(kUnsetSeqNum), set_(false) {}
+
+SequenceNum::SequenceNum(const uint64_t num) : num_(num), set_(true) {}
```
You can just use `std::optional<int>` to represent `num_` and no need for the field `set_`.

rohan-varma committed:
Update on "[c10d] sequence number in process group"


Adds a sequence number class as well as integration with ProcessGroup (nccl and gloo) as part of better debugability.

The main use case is that each ProcessGroup instantiated will have a sequence number initially set by rank 0, and broadcasted to all others. We will increment the number on each collective, thus allowing us to match the numbers appropriately when checking for desynchronization.

This PR just adds the bare-bones integration and verifies sequence numbers are set appropriately at the beginning. The next diff in this stack will increment the sequence number appropriately on collectives.

Note that we initialize sequence numbers in python in `new_group` and `init_process_group` APIs for NCCL and Gloo backends only. The reason we do this instead of the constructor is because the pg constructor in C++ is not a synchronization point, and can result in timeout issues.

Differential Revision: [D27562769](https://our.internmc.facebook.com/intern/diff/D27562769/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27562769/)!

[ghstack-poisoned]

SciPioneer commented on torch/lib/c10d/ProcessGroup.hpp:
```diff
@@ -297,6 +324,8 @@ class ProcessGroup : public torch::CustomClassHolder {
   const int rank_;
   const int size_;
+  // Optional sequence number structure for matching collectives.
+  c10::optional<c10d::SequenceNum> sequenceNum_ = c10::nullopt;
```
Nit: shall we prefer std::optional to c10d::optional?

rohan-varma replied:
I think std::optional is only available in C++17 (https://en.cppreference.com/w/cpp/utility/optional) which pytorch OSS doesn't support yet afaik. The only uses of `std::optional` in `torch/` directory seems to be in fb-only directories.

rohan-varma committed:
Update on "[c10d] sequence number in process group"


Adds a sequence number class as well as integration with ProcessGroup (nccl and gloo) as part of better debugability.

The main use case is that each ProcessGroup instantiated will have a sequence number initially set by rank 0, and broadcasted to all others. We will increment the number on each collective, thus allowing us to match the numbers appropriately when checking for desynchronization.

This PR just adds the bare-bones integration and verifies sequence numbers are set appropriately at the beginning. The next diff in this stack will increment the sequence number appropriately on collectives.

Note that we initialize sequence numbers in python in `new_group` and `init_process_group` APIs for NCCL and Gloo backends only. The reason we do this instead of the constructor is because the pg constructor in C++ is not a synchronization point, and can result in timeout issues.

Differential Revision: [D27562769](https://our.internmc.facebook.com/intern/diff/D27562769/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27562769/)!

[ghstack-poisoned]

zhaojuanmao commented on torch/lib/c10d/test/ProcessGroupNCCLTest.cpp:
```diff
@@ -387,6 +387,31 @@ void testReduceScatter(const std::string& path, int rank, int size) {
+  for (auto& t : threads) {
+    t.join();
+  }
+  EXPECT_EQ(nums.size(), 1);
```
same as above

zhaojuanmao reviewed (commented):
Overall looks good.

I'm wondering whether we can use broadcast instead of store to broadcast seq_num and the n log seq_num in local logs?  as capacity of distributed store file like zeus could be limited. 

what is the advantage to use store over broadcast + local logs?

rohan-varma committed:
Update on "[c10d] sequence number in process group"


Adds a sequence number class as well as integration with ProcessGroup (nccl and gloo) as part of better debugability.

The main use case is that each ProcessGroup instantiated will have a sequence number initially set by rank 0, and broadcasted to all others. We will increment the number on each collective, thus allowing us to match the numbers appropriately when checking for desynchronization.

This PR just adds the bare-bones integration and verifies sequence numbers are set appropriately at the beginning. The next diff in this stack will increment the sequence number appropriately on collectives.

Note that we initialize sequence numbers in python in `new_group` and `init_process_group` APIs for NCCL and Gloo backends only. The reason we do this instead of the constructor is because the pg constructor in C++ is not a synchronization point, and can result in timeout issues.

Differential Revision: [D27562769](https://our.internmc.facebook.com/intern/diff/D27562769/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27562769/)!

[ghstack-poisoned]

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -4905,6 +4905,58 @@ def test_broadcast_coalesced_gloo_cpu(self):
+        seq_num = default_pg._get_sequence_number_for_group()
+        obj_list = [None for _ in range(dist.get_world_size())]
+        dist.all_gather_object(obj_list, seq_num)
+        self.assertEqual(len(set(obj_list)), 1)
```
That should be what this test does, `seq_num` is the number on the rank, and then we all gather it into a `obj_list`. Then if we convert this to a set the set size should be 1 if they are the same across all ranks.

zhaojuanmao replied:
I see, make sense

rohan-varma commented on torch/lib/c10d/test/ProcessGroupGlooTest.cpp:
```diff
@@ -464,6 +464,21 @@ void testMonitoredBarrier(const std::string& path) {
+    auto seqNum = tests[i].getProcessGroup().getSequenceNumberForGroup();
+    nums.insert(seqNum);
+  }
+  EXPECT_EQ(nums.size(), 1);
```
This should be covered by this test, we get the number for each process group and insert it into a set. So the size of the set should be 1 which we check here to ensure they are all the same sequence number. Is this what you meant?

rohan-varma commented:
> Overall looks good.
> 
> I'm wondering whether we can use broadcast instead of store to broadcast seq_num and the n log seq_num in local logs? as capacity of distributed store file like zeus could be limited.
> 
> what is the advantage to use store over broadcast + local logs?

Currently, we only store the initial sequence number in the `Store` and don't plan to store anything else (i.e. subsequent sequence numbers will be local only, and then in debug mode allreduced to verify they are the same).

I think we should be able to use `broadcast` to communicate the first sequence number instead of writing to the store as well, initially I had it done through the store because I was doing this in the process group constructor which is not necessarily a synchronization point. Will refactor to use broadcast here.

rohan-varma commented:
@zhaojuanmao Actually just realized, one difficulty of using `broadcast` especially for pg nccl is that we don't know for sure which GPU to create the tensor to broadcast on. 

For example, when regular user does `ProcessGroupNCCL::broadcast`, we use the GPU that the input tensor is on and create NCCL communicator for that GPU. However, in this case the user is not directly providing us a tensor, and we need to create one on the GPU with something like `rank_ % device_count`. This is similar to what we do for barrier(), which is prone to errors. So it might be best to go with this store-based approach to avoid issues like these?

zhaojuanmao commented:
sounds good to have store broadcast during construction time only, and then log sequence number locally. The cost should be trivial. Thanks!

rohan-varma committed:
Update on "[c10d] sequence number in process group"


Adds a sequence number class as well as integration with ProcessGroup (nccl and gloo) as part of better debugability.

The main use case is that each ProcessGroup instantiated will have a sequence number initially set by rank 0, and broadcasted to all others. We will increment the number on each collective, thus allowing us to match the numbers appropriately when checking for desynchronization.

This PR just adds the bare-bones integration and verifies sequence numbers are set appropriately at the beginning. The next diff in this stack will increment the sequence number appropriately on collectives.

Note that we initialize sequence numbers in python in `new_group` and `init_process_group` APIs for NCCL and Gloo backends only. The reason we do this instead of the constructor is because the pg constructor in C++ is not a synchronization point, and can result in timeout issues.

Differential Revision: [D27562769](https://our.internmc.facebook.com/intern/diff/D27562769/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27562769/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] sequence number in process group"


Adds a sequence number class as well as integration with ProcessGroup (nccl and gloo) as part of better debugability.

The main use case is that each ProcessGroup instantiated will have a sequence number initially set by rank 0, and broadcasted to all others. We will increment the number on each collective, thus allowing us to match the numbers appropriately when checking for desynchronization.

This PR just adds the bare-bones integration and verifies sequence numbers are set appropriately at the beginning. The next diff in this stack will increment the sequence number appropriately on collectives.

Note that we initialize sequence numbers in python in `new_group` and `init_process_group` APIs for NCCL and Gloo backends only. The reason we do this instead of the constructor is because the pg constructor in C++ is not a synchronization point, and can result in timeout issues.

Differential Revision: [D27562769](https://our.internmc.facebook.com/intern/diff/D27562769/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27562769/)!

[ghstack-poisoned]
